man/uman.1
.\"
.\" Copyright (c) 2009-2020 Julien Nadeau Carriere <vedge@csoft.net>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd DECEMBER 4, 2009
.Dt UMAN 1
.Os
.ds vT BSDBuild Reference
.ds oS BSDBuild 2.7
.Sh NAME
.Nm uman
.Nd Uninstalled manual page reader
.Sh SYNOPSIS
.Nm uman
.Op Ar search-string
.Sh DESCRIPTION
The
.Nm
script formats and displays uninstalled manual pages, without using the
system manual.
The
.Ar search-string
argument should be a manual page title (or in the case of library
documentation, the name of a function, structure or class).
.Pp
The
.Dv UMANPATH
environment variable should be defined as a comma-separated list of
directories containing manual page source files such as
.Pa foo.1 ,
and optionally BSDBuild-generated
.Pa .manlink.mk
files.
.Nm
will search the directories recursively.
.Pp
If a match is found,
.Nm
formats and feeds the resulting manual page to the system
.Dv PAGER
(or the default,
.Xr less 1 ) .
.Sh SEE ALSO
.Xr build.common.mk 5 ,
.Xr build.man.mk 5


CHANGELOG.md
# Changelog

All notable changes to BSDBuild will be documented in this file. This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [3.2] - 2020-04-12

### Added

- [**build.prog.mk**](https://bsdbuild.hypertriton.com/man5/build.prog.mk) & [**build.lib.mk**](https://bsdbuild.hypertriton.com/man5/build.lib.mk): Add support for the Ada language. Introduce `$ADA`, `$ADABIND`, `$ADALINK` and `$LINKER_TYPE`. Introduce `${PROG_BUNDLE}`, `${LIB_BUNDLE}` and `gen-bundle.pl` for generating platform-specific application bundles (currently "OSX" or "iOS").
- [**build.www.mk**](https://bsdbuild.hypertriton.com/man5/build.www.mk): Introduce the mlproc(1) multilanguage preprocessor. Define `${MLPROC}`, `${MLPROCFLAGS}`, `${MINIFIER}`, `${MINIFIERFLAGS}` and `${MINIFIERFLAGSCSS}`.
- [**mkconfigure**](https://bsdbuild.hypertriton.com/man1/mkconfigure): New directives: `Ada_option()`, `Ada_bflag()`, `Hdefine_if()`. Export LDFLAGS to Makefile.config. Fix `Ld_Option()`. New directive `Pkgconfig_Module()` for integrated pkg-config module generation. Add `$PKGCONFIG_LIBDIR` setting to simplify the installation of .pc modules. Introduce `--keep-conftest` configure option (preserve output test files).
- **ada**: New test for Ada toolchain.
- **agar-ada**: New test module for Ada bindings to Agar-GUI.
- **agar-ada-core**: New test module for Ada bindings to Agar-Core.
- **agar.types**: New test modules to generate tables of definitions and struct sizes for an installed Agar build. Intended to simplify the implementation of thin and variable-thickness bindings to different languages.
- **cc**: Add support for [cc65](https://cc65.github.io). Add `$CC_COMPILE`. Don't assume the compiler understands "-c" or "-O2".
- **imagemagick**: New test for ImageMagick 6 and 7.
- **lldb**: New test for lldb interface library.
- **tcl**: New test for Tcl. Thanks Chuck!

### Removed

- **freesg**: Module replaced by **agar-sg**.

### Changed

- [**build.lib.mk**](https://bsdbuild.hypertriton.com/man5/build.lib.mk): Disable use of Libtool by default. Generate .la files even when `USE_LIBTOOL=No`. Handle .dylib and .dll files directly.
- [**build.prog.mk**](https://bsdbuild.hypertriton.com/man5/build.prog.mk) & [**build.lib.mk**](https://bsdbuild.hypertriton.com/man5/build.lib.mk): Use ``{LIB,PROG}_PROFILE`` instead of separate .po targets. Merge the "depend" from [**build.dep.mk**](https://bsdbuild.hypertriton.com/man5/build.dep.mk). Include support for cc65 compiler.
- [**mkconfigure**](https://bsdbuild.hypertriton.com/man1/mkconfigure): Make the generated config.log a runnable shell script containing all test code and compilation commands (so individual tests can be reproduced easily by copy/pasting from it). Fix quoting of arguments in config.status.
- [**mkconfigure**](https://bsdbuild.hypertriton.com/man1/mkconfigure): Show influential environment variables in generated "--help" output. Introduce `Register_env_var()` directive. Use `$IFS` instead of sed for path traversals. Add test for `$PATH_SEPARATOR`. Use expr to validate `--with-*` and `--enable-*` arguments. Honor `${EXECSUFFIX}` in `$PATH` searches. Don't invoke config.guess unnecessarily if --build is provided.
- **mkconcurrent** script: Extract and cache make target fragments directly from an installed build.\*.mk so we no longer need to maintain a copy of them in the script.
- **cc**: Define `HAVE_CC_GCC` and `HAVE_CC_CLANG`.
- **x11**: Honor `--x-includes=*` and `--x-libraries=*` configure arguments. If XKB is not found then fallback to libXf86misc.
- **pthreads**: Add subtests `HAVE_PTHREAD_{MUTEX,COND,}_T_POINTER` to determine whether `pthread_mutex_t`, `pthread_cond_t` and `pthread_t` are pointer types or not. Needed for `_Pure_Attribute_If_Unthreaded` in agar.
- Update config.guess to 2015-03-04. Add [FabBSD](https://FabBSD.org).
- Don't imply "all" in the "install" target.
- Updates to the manual pages.

### Fixed

- [**mkconfigure**](https://bsdbuild.hypertriton.com/man1/mkconfigure): Handle string literals containing "," and ";" in directives. Thanks Chuck!
- **freetype**: Check with pkgconfig also. Thanks enthus1ast!
- Don't include non-file args in "nothing to do" evaluation in bundled mkdep script.

## [3.1] - 2015-07-14

### Added

- [**build.man.mk**](https://bsdbuild.hypertriton.com/man5/build.man.mk): Add support for the new [OpenBSD](https://openbsd.org) [mandoc](https://man.openbsd.org/mandoc). Add support for PDF and HTML format output. Introduce new target "lint".
- **csidl**: New test for Windows CSIDL interface.
- **kqueue**: New test for [**kqueue(2)**](https://www.freebsd.org/cgi/man.cgi?kqueue) kernel event notification mechanism.
- **nanosleep**: New test for nanosleep(2) interface.
- **rand48**: New test for rand48() PRNG routines.
- **timerfd**: New test for timerfd on Linux.
- **uim**: New test for uim input method framework.
- **x11**: Add /usr/local/include, /usr/include to fallback test.
- **x11**: Added test for XKB extension (`HAVE_XKB`).
- **xbox**: New test for xbox XDK.

### Removed

- [**mkconfigure**](https://bsdbuild.hypertriton.com/man1/mkconfigure): No longer include the `--cache` support code by default in order to make generated scripts smaller since caching is mostly advantageous on slower systems. Use `CONFIG_CACHE(yes)` to re-enable. Generate **config.status** files.
- **cxx**, **objc**: Remove unnecessary tests for Cygwin, long long and long double.
- **sse**: Remove unnecessary test code.

### Changed

- [**build.prog.mk**](https://bsdbuild.hypertriton.com/man5/build.prog.mk): Allow `${PROG_TYPE}`-dependent LDFLAGS to be defined.
- [**build.lib.mk**](https://bsdbuild.hypertriton.com/man5/build.lib.mk): Updated bundled libtool to 2.4.2. Use the `LIB_{CURRENT,REVISION,AGE}` versioning scheme.

uman.pl
#!%PERL%
#
# Public domain
# 
# Search $UMANPATH for a user-installed manual page and write it to
# the standard output (via groff).
#
# Links can also be specified (uman will search BSDBuild-generated
# .manlinks.mk files for them).
#

if (!defined($ENV{'UMANPATH'})) {
	print "The \$UMANPATH environment variable is not set; see uman(1).\n";
	exit(1);
}

# Search these directories for manual page source.
@SRC_DIRS = split(':', $ENV{'UMANPATH'});

# Formatting engine
#$NROFF = 'nroff -Tascii -mandoc';
$NROFF = 'groff -S -P-h -Wall -mtty-char -man -Tascii -P-c -mandoc';

if (@ARGV < 1) { die "Usage: uman [manpage]"; }
my $query = $ARGV[0];
my $pager = 'less';
if (exists($ENV{'PAGER'})) { $pager = $ENV{'PAGER'}; }

sub ReadPage ($)
{
	my $page = shift;
	system("cat $page |$NROFF $page |$pager");
}

# Search a BSDBuild-generated .manlinks.mk file.
sub SearchManlinksMK ($$)
{
	my $path = shift;
	my $q = shift;

	unless (open(ML, $path)) {
		print STDERR "$path: $!; ignoring\n";
		return;
	}
	foreach $_ (<ML>) {
		if (/^MANLINKS\+=([\w\-]+)\.(\d):([\w\-]+)\.(\d)$/) {
			my $from = $1.'.'.$2;

			if (lc($3) eq $q) {
				return ($from);
			}
		}
	}
	close(ML);
	return (undef);
}

# Recursively search a directory.
sub SearchDir ($$$$)
{
	my $dir = shift;
	my $q = lc(shift);
	my $depth = shift;
	my $maxDepth = shift;

	unless (opendir(DIR, $dir)) {
		return;
	}
	foreach my $ent (readdir(DIR)) {
		my $path = $dir.'/'.$ent;
		if ($ent eq '.manlinks.mk') {
			if ((my $rv = SearchManlinksMK($path, $q))) {
				ReadPage($dir.'/'.$rv);
				exit(0);
			}
			next;
		}
		if ($ent =~ /^\./) { next; }
		if ($ent =~ /^([\w\-\.]+)\.(\d)$/) {
			if ($q eq lc($1) ||
			    $q eq lc($1.'.'.$2)) {
				ReadPage($path);
				exit(0);
			}
		}
		if (-d $path && $depth+1 <= $maxDepth) {
			SearchDir($path, $q, $depth+1, $maxDepth);
		}
	}
	closedir(DIR);
}

foreach my $dir (@SRC_DIRS) {
	SearchDir($dir, $query, 1, 10);
}
